### PR TITLE
Add Base64Fuzzer to the repo

### PR DESCRIPTION
**Pull request by buyaa-n**
Add Base64Fuzzer to the repo

Related to https://github.com/dotnet/runtime/pull/102364#pullrequestreview-2072505599

### CODE REVIEW

File: src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+{
+    internal class Base64Fuzzer : IFuzzer
+    {
+        public string[] TargetAssemblies => [];
+
+        public string[] TargetCoreLibPrefixes => ["System.Buffers.Text"];
+
+        public void FuzzTarget(ReadOnlySpan<byte> bytes)
+        {
+            using var inputPoisoned = PooledBoundedMemory<byte>.Rent(bytes, PoisonPagePlacement.After);
```
Comment (bartonjs) [inline src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:18]: Inappropriate use of `var`

Comment (bartonjs) [reply src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:18]: All of the uses of var that I see in this file so far are inappropriate.

Comment (buyaa-n) [reply src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:18]: Just followed existing fuzzers style, anyway, will change it https://github.com/dotnet/runtime/blob/c286a8efb3c0d534557dde476d5b54b15c0aeaed/src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/UTF8Fuzzer.cs#L34-L35

**Comment by MihaZupan** [comment]
@MihuBot fuzz base64

File: src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+                    Assert.Equal(OperationStatus.Done, status);
+                    Assert.Equal(bytes.Length, bytesDecoded);
+                    Assert.Equal(bytesEncoded, bytesRead);
+                    Assert.SequenceEqual(bytes, decoderDest.Slice(0, bytesDecoded));
+                }
+                else
+                {
+                    Assert.Equal(OperationStatus.NeedMoreData, status);
+                    Assert.Equal(true, maxEncodedLength >= bytesEncoded);
```
Comment (MihaZupan) [inline src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:64]: Should this assert the exact expected value instead?

E.g. `Assert.Equal(input.Length / 3 * 4, bytesEncoded)` (or whichever formula applies here)

File: src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                    if (decodeInput.Length % 4 == 0)
+                    {
+                        Assert.Equal(OperationStatus.Done, status);
+                        Assert.Equal(bytesConsumed, bytesDecoded);
+                        Assert.Equal(bytesEncoded, bytesRead);
+                    }
+                    else
+                    {
+                        Assert.Equal(OperationStatus.NeedMoreData, status);
+                        Assert.SequenceEqual(bytes.Slice(0, bytesDecoded), decoderDest.Slice(0, bytesDecoded));
```
Comment (MihaZupan) [inline src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:77]: This assert looks like it would apply to the `if` block as well

**Review by MihaZupan — APPROVED** [review]
The fuzzing part LGTM.

The System.Buffers area may want to check if this is covering all the paths we care about for Base64.

File: src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:64]: The formula does not return exact encoded value, exact encoded value can be equal to `maxEncodedLength`, `maxEncodedLength  - 1`, or `maxEncodedLength - 2` depending on padding count https://github.com/dotnet/runtime/blob/db5e98d64641784213f5e88b3d1523c7ef47c01d/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Encoder.cs#L174

I can update the conditions with ORs

Comment (MihaZupan) [reply src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:64]: I would expect that with `OperationStatus.NeedMoreData`, the remainder of the input shouldn't be consumed.
After an `EncodeToUtf8` call, `bytesConsumed` should be divisible by 3, meaning that `written` would be `consumed / 3 * 4`.
Is that not the case?

Comment (buyaa-n) [reply src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/Base64Fuzzer.cs:64]: Ah I see you meant for this exact case not in general, yes, you are right

**Comment by buyaa-n** [comment]
> The System.Buffers area may want to check if this is covering all the paths we care about for Base64.

Well, Base64 doesn't have so many APIs, and this PR covers all those APIs. There could be scenario where whitespaces involved in decoding, but I am not sure how to tell the fuzzer to generate valid base64 bytes with whitespaces, looks there is no way except insert white spaces manually after encoding.

**Pull request merged**